keps/sig-cluster-lifecycle/draft-20180130-kubeadm-join-master.md

```diff
@@ -0,0 +1,430 @@
+  user stories for creating an highly available Kubernetes cluster, but instead
+  focuses on:
+
+  * Defining a generic and extensible flow for bootstrapping an HA cluster, the
+    `kubeadm join --master` workflow.
+  * Providing a solution *only* for one, well defined user story. see
+    [User Stories](#user-stories) and [Non-goals](#non-goals).
+
+* Provide support for a dynamic bootstrap flow
+
+  At the time a user is running `kubeadm init`, s/he/an operator might not know what
+  the cluster setup will look like eventually. For instance, the user may start with
+  only one master + n nodes,  and then add further master nodes with `kubeadm join --master`
+  or add more worker nodes with `kubeadm join` (in any order).
+
+* Enable higher-level tools integration
+
+  We expect higher-level and more tailored tooling to be built on top of kubeadm,
+  and ideally, using kubeadm as the basis of all deployments will make it easier
+  to create conformant cluster.
+
+  Accordingly, the `kubeadm join --master` workflow should provide support for
+  the following operational practices used by higher level tools:
+
+  * Parallel node creation
+
+    Higher-level tools could create nodes in parallel (both masters and workers)
+    for reducing the overall cluster startup time.
+    `kubeadm join --master` should support natively this practice without requiring
+    the implementation of any synchronization mechanics by higher-level tools
+    (provided there has been an initial seeded master).
+
+  * Replace reconciliation strategies
+
+    Especially in case of cloud deployments, higher-level automation tools could
+    decide for any reason to replace an existing nodes with new ones (instead of apply
+    changes in-place to existing nodes).
+    `kubeadm join --master` will support this practice by making easier to replace
+    existing master nodes with new ones.
+
+### Non-goals
+
+* By design, kubeadm cares only about bootstrapping, not about provisioning machines.
+  Likewise, installing various nice-to-have addons, like the Kubernetes Dashboard,
+  monitoring solutions, and cloud-specific addons, is not in scope.
+
+* This proposal doesn't include a solution for etcd cluster management.
+
+  NB 1. At the time of writing, running etcd outside kubernetes is the recommended
+  approach from CoreOS (see discussion in [kubeadm office hours](https://goo.gl/fjyeqo)).
+
+  NB 2. Nothing in this proposal should prevent users to run master nodes components
+  and etcd on the same machines; however users should be aware that this will
+  introduce limitations for strategies like parallel node creation and In-place
+  vs. Replace reconciliation.
+
+  NB 3. Nothing in this proposal should prevent kubeadm to implement in future a
+  solution for provisioning an etcd cluster based on static pods/pods.
+
+* This proposal doesn't include a solution for API server load balancing.
```

Comment:
This line seems out of place with the "list of notes" structure here. Would it fit better up near line 124 where it's mentioned that this proposal doesn't include a solution for etcd? Or maybe as part of `NB 4`?